### PR TITLE
Python 3: Fix certs/x509/TLS1/SSLv2/TLS1.3

### PR DESCRIPTION
**Pull request by gpotter2**
- fixes cert.uts
- fixes x509.uts
- fixes tls.uts (except one test: it seems that cryptography API has changed, not sure how to fix it)
- fixes sslv2.uts
- fixes tls13.uts

### CODE REVIEW

**Comment by gpotter2** [comment]
@mtury Hi, I'd really like to have your review for one ^^
Tell me if it breaks any (future?) things !

**Comment by gpotter2** [comment]
Ready to be reviewed.
@guedou @p-l-

File: scapy/layers/tls/session.py [hunk 1]
```diff
@@ -89,7 +90,7 @@ def __init__(self,
         self.cipher = ciphersuite.cipher_alg()
         self.hash = ciphersuite.hash_alg()
 
-        if tls_version > 0x0200:
+        if tls_version and tls_version > 0x0200:
```
Comment (mtury) [inline scapy/layers/tls/session.py]: Same comment.

File: scapy/layers/tls/session.py [hunk 2]
```diff
@@ -100,7 +101,7 @@ def __init__(self,
             self.hmac = ciphersuite.hmac_alg()          # should be Hmac_NULL
             self.mac_len = self.hash.hash_len
 
-        if tls_version >= 0x0304:
+        if tls_version and tls_version >= 0x0304:
```
Comment (mtury) [inline scapy/layers/tls/session.py]: Same comment.

File: test/sslv2.uts [hunk 3]
```diff
@@ -234,29 +234,29 @@ t.msg[0].challenge == challenge
...
 = Building SSLv2 packets - Error within clear record
 t = SSLv2(msg=SSLv2Error(code='no_cipher'))
-str(t) == b'\x80\x03\x00\x00\x01'
+raw(t) == b'\x80\x03\x00\x00\x01'
 
 = Building SSLv2 packets - ClientHello with automatic length computation
 ch_pkt = SSLv2ClientHello()
 ch_pkt.msgtype = 'client_hello'
-ch_pkt.version = 'SSLv2'
+ch_pkt.version = 0x002
```
Comment (mtury) [inline test/sslv2.uts]: I'd rather see 0x0002

Comment (gpotter2) [reply test/sslv2.uts]: Right

**Comment by gpotter2** [comment]
@mtury This is indeed a messy change, but it results from an existing bug, undiscovered because of Python 2 basics.
- On python 2 ONLY, we had this
```
>>> None < 0
True
>>> None > 1
False
```
- On python 3, with every thing (apart from this fixed), i get errors for two tests only, of `sslv2.uts`:
### Reading SSLv2 session - ClientMasterKey with unknown server key 
```
>>> t_enc = SSLv2(mk)
Traceback (most recent call last):
  File "<input>", line 2, in <module>
  File "Z:\Coding\github\scapy\scapy\base_classes.py", line 231, in __call__
    i.__init__(*args, **kargs)
  File "Z:\Coding\github\scapy\scapy\layers\tls\record_sslv2.py", line 81, in __init__
    super(SSLv2, self).__init__(*args, **kargs)
  File "Z:\Coding\github\scapy\scapy\layers\tls\record.py", line 262, in __init__
    super(TLS, self).__init__(*args, **kargs)
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 828, in __init__
    **fields)
  File "Z:\Coding\github\scapy\scapy\packet.py", line 127, in __init__
    self.dissect(_pkt)
  File "Z:\Coding\github\scapy\scapy\packet.py", line 734, in dissect
    s = self.do_dissect(s)
  File "Z:\Coding\github\scapy\scapy\packet.py", line 697, in do_dissect
    s, fval = f.getfield(self, s)
  File "Z:\Coding\github\scapy\scapy\layers\tls\record.py", line 144, in getfield
    p.post_dissection_tls_session_update(raw_msg)
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 850, in post_dissection_tls_session_update
    self.tls_session_update(msg_str)
  File "Z:\Coding\github\scapy\scapy\layers\tls\handshake_sslv2.py", line 360, in tls_session_update
    tls_version=tls_version)
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 283, in __init__
    connState.__init__(self, read_or_write="write", **kargs)
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 86, in __init__
    self.ciphersuite = ciphersuite(tls_version=tls_version)
  File "Z:\Coding\github\scapy\scapy\layers\tls\crypto\suites.py", line 141, in __init__
    if tls_version <= 0x301:
TypeError: '<=' not supported between instances of 'NoneType' and 'int'
```
and
### Building SSLv2 packets - ClientMasterKey context linking 
```
>>> mk = SSLv2ClientMasterKey(cipher='SSL_CK_DES_192_EDE3_CBC_WITH_MD5', clearkey=b'\xff'*19, encryptedkey=b'\0'*256, decryptedkey=b'\xaa'*5, keyarg=b'\x01'*8)
>>> t = SSLv2(msg=mk)
>>> t.tls_session.sslv2_connection_id = b'\xba'*16
>>> t.tls_session.sslv2_challenge = b'\x42'*16
>>> t.str_stateful()
Traceback (most recent call last):
  File "<input>", line 2, in <module>
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 863, in str_stateful
    return super(_GenericTLSSessionInheritance, self).__bytes__()
  File "Z:\Coding\github\scapy\scapy\packet.py", line 338, in __bytes__
    return self.build()
  File "Z:\Coding\github\scapy\scapy\packet.py", line 436, in build
    p = self.do_build()
  File "Z:\Coding\github\scapy\scapy\packet.py", line 418, in do_build
    pkt = self.self_build()
  File "Z:\Coding\github\scapy\scapy\packet.py", line 399, in self_build
    p = f.addfield(self, p, val)
  File "Z:\Coding\github\scapy\scapy\layers\tls\record_sslv2.py", line 62, in addfield
    res += self.i2m(pkt, p)
  File "Z:\Coding\github\scapy\scapy\layers\tls\record_sslv2.py", line 52, in i2m
    p.post_build_tls_session_update(cur)
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 847, in post_build_tls_session_update
    self.tls_session_update(msg_str)
  File "Z:\Coding\github\scapy\scapy\layers\tls\handshake_sslv2.py", line 360, in tls_session_update
    tls_version=tls_version)
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 283, in __init__
    connState.__init__(self, read_or_write="write", **kargs)
  File "Z:\Coding\github\scapy\scapy\layers\tls\session.py", line 86, in __init__
    self.ciphersuite = ciphersuite(tls_version=tls_version)
  File "Z:\Coding\github\scapy\scapy\layers\tls\crypto\suites.py", line 141, in __init__
    if tls_version <= 0x301:
TypeError: '<=' not supported between instances of 'NoneType' and 'int'
```

This is sounds like an undiscovered bug that was forgotten (while creating the tests ?)

File: scapy/layers/tls/crypto/prf.py [hunk 4]
```diff
@@ -178,7 +179,7 @@ def __init__(self, hash_name="SHA256", tls_version=0x0303):
         self.tls_version = tls_version
         self.hash_name = hash_name
 
-        if tls_version < 0x0300:            # SSLv2
+        if (not tls_version) or tls_version < 0x0300:  # SSLv2
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/layers/tls/crypto/prf.py]: See my comment below

File: scapy/layers/tls/session.py [hunk 5]
```diff
@@ -412,7 +413,7 @@ def __init__(self,
         self.advertised_tls_version = 0x0303
 
         # The agreed-upon TLS version found in the ServerHello.
-        self.tls_version = None
+        self.tls_version = 0
```
Comment (gpotter2) [inline scapy/layers/tls/session.py]: This will fix the `None > 0x0303` Bug.

Comment (p-l-) [reply scapy/layers/tls/session.py]: Sorry, but I think I prefer the first way you had fixed it. `None` has a special meaning. @mtury what do you think?

Comment (gpotter2) [reply scapy/layers/tls/session.py]: I could set -1 if it's clearer

Comment (p-l-) [reply scapy/layers/tls/session.py]: I don't think it's clearer. `None` is clearer. The problem is you compare a value, with a "non-value". If you use `0` (or even `-1`), that looks like a value, and you use the hack that your non-value compares with a value (that's what happened with `None` compared to integers in Python 2).

**Comment by mtury** [comment]
PR #865 should fix the issue, so that we do not need your proposed `not tls_version` conditions. Thanks for catching it!

**Pull request merged**